File: C++/006_ZigZag Conversion.cpp
class Solution {
public:
    string convert(string s, int nRows) {
        if (1 == nRows)
            return s;
        string res = s;
        int i, j, k, l = (nRows << 1) - 2;
        int len = (int)s.length();
        for (i = j = 0; i < len; i += l)
            res[j++] = s[i];
        for (k = 1; k < nRows - 1; k++) {
            int step = l - (k << 1);
            for (i = k; i < len; i += l) {
                res[j++] = s[i];
                if (i + step < len)
                    res[j++] = s[i + step];
            }
        }
        for (i = nRows - 1; i < len; i += l)
            res[j++] = s[i];
        return res;
    }
};

File: Python/006_ZigZag Conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if numRows <= 1:
            return s
        up, l, res = 2 * numRows - 2, len(s), ''
        for i in range(0, l, up):
            res += s[i]
        for i in range(1, numRows - 1):
            for j in range(i, l, up):
                res += s[j]
                if j + up - 2 * i < l:
                    res += s[j + up - 2 * i]
        for i in range(numRows - 1, l, up):
            res += s[i]
        return res
        